chips/stm32f4xx/src/flash.rs

```diff
@@ -0,0 +1,667 @@
+    }
+}
+
+/// Tests for the STM32F4xx flash driver.
+///
+/// If any contributions are made to this driver, it is highly recommended to run these tests to
+/// ensure that everything still works as expected. The tests are chip agnostic. They can be run
+/// on any STM32F4 chips.
+///
+/// # Usage
+///
+/// First, the flash module must be imported:
+///
+/// ```rust,ignore
+/// // Change this line depending on the chip the board is using
+/// use stm32f429zi::flash;
+/// ```
+///
+/// Then, get a reference to the peripheral:
+///
+/// ```rust,ignore
+/// // Inside the board main.rs
+/// let flash = &peripherals.stm32f4.flash;
+/// ```
+///
+/// To run all tests:
+///
+/// ```rust,ignore
+/// flash::tests::run_all(flash);
+/// ```
+///
+/// The following output should be printed:
+///
+/// ```text
+/// ===============================================
+/// Testing setting flash latency...
+///
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// Testing number of wait cycles based on the system frequency...
+/// Finished testing number of wait cycles based on the system clock frequency. Everything is alright!
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+///
+///
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// Testing setting flash latency...
+/// Finished testing setting flash latency. Everything is alright!
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+///
+/// Finished testing flash. Everything is alright!
+/// ===============================================
+/// ```
+///
+/// To run individual tests, see the functions in this module.
+///
+/// # Errors
+///
+/// In case of any errors, open an issue ticket at <https://github.com/tock/tock>. Please provide
+/// the output of the test execution.
+pub mod tests {
+    use super::*;
+
+    const HSI_FREQUENCY_MHZ: usize = 16;
+    const AHB_ETHERNET_MINIMUM_FREQUENCY_MHZ: usize = 25;
+    // Different chips have different maximum values for APB1
+    const APB1_MAX_FREQUENCY_MHZ_1: usize = 42;
+    const APB1_MAX_FREQUENCY_MHZ_2: usize = 45;
+    const APB1_MAX_FREQUENCY_MHZ_3: usize = 50;
+    // Different chips have different maximum values for APB2
+    const APB2_MAX_FREQUENCY_MHZ_1: usize = 84;
+    #[cfg(not(feature = "stm32f401"))] // Not needed for this chip model
+    const APB2_MAX_FREQUENCY_MHZ_2: usize = 90;
+    #[cfg(not(feature = "stm32f401"))] // Not needed for this chip model
+    const APB2_MAX_FREQUENCY_MHZ_3: usize = 100;
+    // Many STM32F4 chips allow a maximum frequency of 168MHz and some of them 180MHz if overdrive
+    // is turned on
+    #[cfg(not(feature = "stm32f401"))] // Not needed for this chip model
+    const SYS_MAX_FREQUENCY_NO_OVERDRIVE_MHZ: usize = 168;
+    #[cfg(not(feature = "stm32f401"))] // Not needed for this chip model
+    const SYS_MAX_FREQUENCY_OVERDRIVE_MHZ: usize = 180;
+    // Default PLL frequency
+    #[cfg(not(feature = "stm32f401"))] // Not needed for this chip model
+    const PLL_FREQUENCY_MHZ: usize = 96;
```

Comment:
Writing chip-agnostic tests was also a problem. Because of different requirements such as maximum system clock frequency, some tests value don't make any sense on some chip models. It is true that a unified test version could have been created, but a compromise has to be made:

+ Either reduce the test domain to a small subset common to all chips
+ Or introduce non-sense test cases for some chips.

I decided to create coeherent tests without making any compromise. The downside is the same as for writing the driver itself: conditional compilation had to be used.

Again, I would like to hear your feedback on this!